File: go/300.500/429.NAryTreeLevelOrderTraversal.go
package main

import "fmt"

type Node struct {
	Val      int
	Children []*Node
}

func levelOrder(root *Node) [][]int {
	if root == nil {
		return [][]int{}
	}
	ret := [][]int{}
	que := []*Node{root}
	for len(que) > 0 {
		nque := []*Node{}
		ret1 := []int{}
		for _, c := range que {
			ret1 = append(ret1, c.Val)
			for _, x := range c.Children {
				nque = append(nque, x)
			}
		}
		ret = append(ret, ret1)
		que = nque
	}
	return ret
}

func main() {
	head := &Node{Val: 1}
	head.Children = append(head.Children, &Node{Val: 3})
	head.Children = append(head.Children, &Node{Val: 2})
	head.Children = append(head.Children, &Node{Val: 4})
	head.Children[0].Children = append(head.Children[0].Children, &Node{Val: 5})
	head.Children[0].Children = append(head.Children[0].Children, &Node{Val: 6})
	ret := levelOrder(head)
	fmt.Println(ret)

	head = &Node{Val: 1}
	head.Children = append(head.Children, &Node{Val: 2})
	head.Children = append(head.Children, &Node{Val: 3})
	head.Children = append(head.Children, &Node{Val: 4})
	head.Children = append(head.Children, &Node{Val: 5})
	head.Children[1].Children = append(head.Children[1].Children, &Node{Val: 6})
	head.Children[1].Children = append(head.Children[1].Children, &Node{Val: 7})
	head.Children[2].Children = append(head.Children[2].Children, &Node{Val: 8})
	head.Children[3].Children = append(head.Children[3].Children, &Node{Val: 9})
	head.Children[3].Children = append(head.Children[3].Children, &Node{Val: 10})
	head.Children[1].Children[1].Children = append(head.Children[1].Children[1].Children, &Node{Val: 11})
	head.Children[1].Children[1].Children[0].Children = append(head.Children[1].Children[1].Children[0].Children, &Node{Val: 14})
	head.Children[2].Children[0].Children = append(head.Children[2].Children[0].Children, &Node{Val: 12})
	head.Children[3].Children[0].Children = append(head.Children[3].Children[0].Children, &Node{Val: 13})
	ret = levelOrder(head)
	fmt.Println(ret)
}

File: 300.500/429.NAryTreeLevelOrderTraversal.cpp
#include <iostream>
#include <queue>
#include <vector>
using namespace std;

class Node {
public:
    int val;
    vector<Node*> children;

    Node() {}

    Node(int _val) {
        val = _val;
    }

    Node(int _val, vector<Node*> _children) {
        val = _val;
        children = _children;
    }
};

class NAryTreeLevelOrderTraversal{
public:
    vector<vector<int>> levelOrder(Node* root) {
        vector<vector<int>> ret;
        if (!root)
            return ret;
        queue<Node*> q;
        q.push(root);
        while (!q.empty())
        {
            int size = q.size();
            ret.resize(ret.size() + 1);
            for (int i = 0; i < size;++i) {
                Node* c = q.front();
                q.pop();
                ret.back().push_back(c->val);
                for (auto* n : c->children) {
                    q.push(n);
                }
            }
        }
        
        return ret;
    }
};

void Delete(Node* head) {
    if (head->children.size() > 0) {
        for (auto* c : head->children) {
            Delete(c);
        }
    }
    delete head;
}

void Print(vector<vector<int>>& ret) {
    cout << "[";
    for (auto& cs : ret) {
        cout << "[";
        for (int c : cs) {
            cout << c << ",";
        }
        cout << "],";
    }
    cout << "]" << endl;
}

int main(int argc, char const* argv[]) {
    NAryTreeLevelOrderTraversal cls;
    Node* head = new Node(1);
    head->children.push_back(new Node(3));
    head->children.push_back(new Node(2));
    head->children.push_back(new Node(4));
    head->children[0]->children.push_back(new Node(5));
    head->children[0]->children.push_back(new Node(6));
    auto& ret = cls.levelOrder(head);
    Print(ret);
    Delete(head);
    
    head = new Node(1);
    head->children.push_back(new Node(2));
    head->children.push_back(new Node(3));
    head->children.push_back(new Node(4));
    head->children.push_back(new Node(5));
    head->children[1]->children.push_back(new Node(6));
    head->children[1]->children.push_back(new Node(7));
    head->children[2]->children.push_back(new Node(8));
    head->children[3]->children.push_back(new Node(9));
    head->children[3]->children.push_back(new Node(10));
    head->children[1]->children[1]->children.push_back(new Node(11));
    head->children[2]->children[0]->children.push_back(new Node(12));
    head->children[3]->children[0]->children.push_back(new Node(13));
    head->children[1]->children[1]->children[0]->children.push_back(new Node(14));
    auto& ret1 = cls.levelOrder(head);
    Print(ret1);
    Delete(head);
    return 0;
}
